### PR TITLE
Created Jobs framework

### PR DESCRIPTION
**Pull request by RageCage64**
Added new framework for creating and running backend jobs. These will be
used by developers to quickly scaffold data for their environments, or
to do important tasks on a running server.

Also fixed a weird data validation thing I discovered while creating and
testing this feature.

### CODE REVIEW

**Comment by RageCage64** [comment]
Screenshot for running the `user` job as an example. (Unrelated: demonstrates that the User validation regex is probably zonked)
image

**Comment by zysim** [comment]
This is so cool

File: LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		{
+			Name = name,
+			Slug = slug,
+		};
+		Context.Leaderboards.Add(newLeaderboard);
+		await Context.SaveChangesAsync();
+		long? newLeaderboardId = Context.Leaderboards.Where(l => l.Name == name).FirstOrDefault()?.Id;
+		if (newLeaderboardId is null)
+		{
+			throw new Exception("Unexpected error creating leaderboard.");
```
Comment (zysim) [inline LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs:35]: You know if - assuming this happens - that the DB row would still exist? I'm wondering if we'd need to worry about having to clean invalid rows if this ever happens. Same goes for L44, for creating users.

In fact; _would_ this ever even happen? Cuz if so, we should also add this check in `ScaffoldUser.Run()` too

File: LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	private ApplicationContext Context;
+
+	public ScaffoldUser(ApplicationContext context)
+	{
+		Context = context;
+	}
+
+	public string CommandName => "user";
+	public string Description => "Create a new user.";
+	public override string ToString() => $"{CommandName}: {Description}";
```
Comment (zysim) [inline LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs:19]: Could we add these fields to the interface instead? Or would (for lack of a better term) adding `ToString` mean that `IJob` would have to be an abstract class? Combination request and query about basic C# here

File: LeaderboardBackend.Jobs/Core/Options.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+				Console.WriteLine(invalidMessage);
+			}
+		} while (!validator(input));
+
+		return input ?? "";
+	}
+
+	public static bool YesOrNo(string question)
+	{
+		Console.Write($"{question} (y to say yes): ");
```
Comment (zysim) [inline LeaderboardBackend.Jobs/Core/Options.cs:33]: We could do the conventional `y/[N]` thing instead of `y to say yes`

File: LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+	async Task IJob.Run()
+	{
+		User newUser = Options.User();
+		newUser.Admin = Options.YesOrNo("Will this user be an admin?");
+		Context.Users.Add(newUser);
+		await Context.SaveChangesAsync();
+	}
```
Comment (zysim) [inline LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs:27]: Refer to comment in `ScaffoldLeaderboardAndMod.cs`

File: LeaderboardBackend/Models/Requests/CategoryRequests.cs [hunk 5]
```diff
@@ -4,7 +4,7 @@
 namespace LeaderboardBackend.Models.Requests;
 
 /// <summary>Request object sent when creating a Category.</summary>
-public record CreateCategoryRequest
+public class CreateCategoryRequest
```
Comment (zysim) [inline LeaderboardBackend/Models/Requests/CategoryRequests.cs:7]: Why was this changed to a class?

**Review by zysim — CHANGES_REQUESTED** [review]

File: LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs (hunk 2 above)
Comment (RageCage64) [reply LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs:19]: These fields are on the interface

File: LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs (hunk 1 above)
Comment (RageCage64) [reply LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs:35]: This wouldn't happen I'm just appeasing null reference type warnings

File: LeaderboardBackend.Jobs/Core/Options.cs (hunk 3 above)
Comment (RageCage64) [reply LeaderboardBackend.Jobs/Core/Options.cs:33]: Makes the code more annoying to check explicitly for `y/n` and I figured it wasn't a big enough deal

File: LeaderboardBackend/Models/Requests/CategoryRequests.cs (hunk 5 above)
Comment (RageCage64) [reply LeaderboardBackend/Models/Requests/CategoryRequests.cs]: accident

File: LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs (hunk 2 above)
Comment (RageCage64) [reply LeaderboardBackend.Jobs/Jobs/ScaffoldUser.cs:19]: Changed the Run() to implement implicitly so they are all consistent

File: LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs (hunk 1 above)
Comment (zysim) [reply LeaderboardBackend.Jobs/Jobs/ScaffoldLeaderboardAndMod.cs:35]: Mmm

File: LeaderboardBackend.Jobs/Core/Options.cs (hunk 3 above)
Comment (zysim) [reply LeaderboardBackend.Jobs/Core/Options.cs:33]: Oh nah cuz `[N]` implies replying with nothing means a rejection, and since we only check for ys, it works out. It's a nitpick, of course

**Review by zysim — DISMISSED** [review]
Depends if you wanna resolve my last remaining comment. I'm happy to get this merged in.

File: LeaderboardBackend.Jobs/Core/Options.cs (hunk 3 above)
Comment (RageCage64) [reply LeaderboardBackend.Jobs/Core/Options.cs:33]: True, I didn't think of it that way. I'll do that.
